File: kthSmallestInMatrix.cpp
// Time Complexity : O(mn)logk; m rows n cols 
// Space Complexity : O(k)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// 1. Kth smallest element of n elements is same as largest of k smallest elements
// 2. Create a max-heap and add first k elements of matrix. Later add new elements only if its smaller than top of heap
// 3. Return top of heap as final result

class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
       if(matrix.size()==0)
            return -1; 
       // largest of k smallest elements
        priority_queue<int> pq;
        for(int i=0; i<matrix.size(); i++){
            for(int j=0; j<matrix[0].size(); j++){
                if(pq.size() < k)
                    pq.push(matrix[i][j]);
                else if(matrix[i][j] < pq.top()){
                    pq.pop();
                    pq.push(matrix[i][j]);
                }
            }
        }
        return pq.pop();
    }
};

File: KthSmallestinMatrix.java
import java.util.PriorityQueue;

// Time Complexity : nlog(n)
// Space Complexity :  O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

public class KthSmallestinMatrix {

    public int kthSmallest(int[][] matrix, int k) {

        if(matrix.length==0 || matrix[0].length==0)
            return 0;
        
		PriorityQueue<Integer> pq= new PriorityQueue<>((a,b) -> b-a);
        
		for(int i =0; i< matrix.length; i++){
			for(int j =0; j< matrix[0].length; j++){
				pq.add(matrix[i][j]);
				if(pq.size()>k)
					pq.remove();
			}
		}
		return pq.peek();
    }
    
}
